### PR TITLE
Updated Nodejs. Updated Typescript. Added ESLint. Added Prettier. Fix…

### PR DESCRIPTION
**Pull request by mgordel**
Updates and cleanup (#293)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,7 +5,7 @@ The JS/TS port of [yapapi, Golem's Python API](https://github.com/golemfactory/y
 
 ## Prerequisites
 
-- [node](https://nodejs.org/en/) >= v12.13.0
+- [node](https://nodejs.org/en/) >= v14.17.0
```
Comment (shadeofblue) [inline README.md:8]: https://github.com/golemfactory/yagna-docs/issues/110

File: examples/hello-world/hello.js [hunk 2]
```diff
@@ -11,18 +16,15 @@ async function main() {
...
     }
   }
 
-  await asyncWith(
-    new Executor({ task_package: package, budget: "1.0", subnet_tag: "devnet-beta" }),
-    async (executor) => {
-      for await (let completed of executor.submit(worker, tasks)) {
-        console.log(completed.result().stdout);
-      }
+  await asyncWith(new Executor({ task_package, budget: "1.0", subnet_tag: "devnet-beta" }), async (executor) => {
```
Comment (shadeofblue) [inline examples/hello-world/hello.js:23]: actually, I'd recommend against this kind of change, in my opinion it decreases the code's readability...

File: examples/low-level-api/listOffers.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@ import Bluebird from "bluebird";
 import dayjs from "dayjs";
 import utc from "dayjs/plugin/utc";
 import { props as yp, rest, utils } from "yajsapi";
-import { Subscription } from "yajsapi/dist/rest/market";
+import { Subscription } from "../../yajsapi/rest/market";
```
Comment (shadeofblue) [inline examples/low-level-api/listOffers.ts:5]: hmm... doesn't it mean that we're depending on the source version here? I guess it would be better to use the installed npm package if possible...

File: examples/low-level-api/listOffers.ts [hunk 4]
```diff
@@ -12,22 +12,22 @@ const { asyncWith, CancellationToken } = utils;
 const cancellationToken = new CancellationToken();
 
 async function list_offers(conf: rest.Configuration, subnetTag: string): Promise<void> {
-  let client = await conf.market();
-  let market_api = new Market(client);
-  let dbuild = new yp.DemandBuilder();
+  const client = await conf.market();
+  const market_api = new Market(client);
+  const dbuild = new yp.DemandBuilder();
```
Comment (shadeofblue) [inline examples/low-level-api/listOffers.ts:17]: okay, here's where my js is not sharp enough -> doesn't `const` suggest that the variable will not be modified? I guess the meaning is obvious when it comes to immutable values... but here it's a mutable and while indeed, we won't be assigning to the `dbuild`, I guess its value _will_ get modified, won't it? ... 

that doesn't feel like a constant to me then... but if it's still okay by the js lingo, I'm okay

File: examples/unsubscribe.ts [hunk 5]
```diff
@@ -1,16 +1,16 @@
 import dayjs from "dayjs";
 import utc from "dayjs/plugin/utc";
 
-import { rest } from "yajsapi";
+import { rest } from "../yajsapi";
```
Comment (shadeofblue) [inline examples/unsubscribe.ts:4]: again, doesn't it mean that we're depending on the source version of yajsapi?

File: tests/goth/test_run_blender.py [hunk 6]
```diff
@@ -103,7 +103,7 @@ async def test_run_blender(
         requestor = runner.get_probes(probe_type=RequestorProbe)[0]
 
         async with requestor.run_command_on_host(
-            f"node {blender_path} --subnet-tag goth",
+            f"npm run --prefix {examples_dir} ts:blender -- -d --subnet-tag goth",
```
Comment (shadeofblue) [inline tests/goth/test_run_blender.py:106]: just out of curiosity - what's the difference between these two invocations?

File: tests/utils/promisify.ts [hunk 7]
```diff
@@ -18,5 +18,5 @@ test("promisify reject given fn", async (t) => {
   }
 
   const error = await t.throwsAsync(promisify(wantCb)(21));
-  t.is(error.message, "i am an error");
+  t.is(error!.message, "i am an error");
```
Comment (shadeofblue) [inline tests/utils/promisify.ts:21]: could you explain this syntax? (again, I'm mostly curious)

File: yajsapi/executor/ctx.ts [hunk 8]
```diff
@@ -20,55 +22,59 @@ export class CommandContainer {
...
           }
-          let idx = self._commands.length;
+          const idx = self._commands.length;
           self._commands.push({ [name]: _arguments });
           return idx;
         };
         return new Proxy(newFunction, {
           apply: function (target, thisArg, argumentsList) {
-            return target.apply(thisArg, argumentsList);
+            return target.apply(thisArg, argumentsList as []);
```
Comment (shadeofblue) [inline yajsapi/executor/ctx.ts:44]: is this a ts's way of saying "argumentsList" is an array?

File: yajsapi/executor/ctx.ts [hunk 9]
```diff
@@ -233,11 +238,9 @@ class _RecvFile extends Work {
 
   async post(): Promise<void> {
     if (!this._dst_slot) throw "_RecvFile post without prepare";
-    if (this._emitter)
-      this._emitter(new events.DownloadStarted({ path: this._src_path }));
+    if (this._emitter) this._emitter(new events.DownloadStarted({ path: this._src_path }));
```
Comment (shadeofblue) [inline yajsapi/executor/ctx.ts:241]: wouldn't this be more readable: 
```suggestion
    if (this._emitter) { 
        this._emitter(new events.DownloadStarted({ path: this._src_path }));
    }
```
?

File: yajsapi/executor/ctx.ts [hunk 10]
```diff
@@ -233,11 +238,9 @@ class _RecvFile extends Work {
...
     await this._dst_slot.download_file(this._dst_path);
-    if (this._emitter)
-      this._emitter(new events.DownloadFinished({ path: this._dst_path }));
+    if (this._emitter) this._emitter(new events.DownloadFinished({ path: this._dst_path }));
```
Comment (shadeofblue) [inline yajsapi/executor/ctx.ts:243]: same here...

File: yajsapi/executor/ctx.ts [hunk 11]
```diff
@@ -293,13 +296,13 @@ export class WorkContext {
   private _started: boolean;
   private _emitter: Callable<[StorageEvent], void> | null;
   private _network_node?: NetworkNode;
-  public provider_info: { provider_id: string, provider_name: string };
+  public provider_info: { provider_id: string; provider_name: string };
```
Comment (shadeofblue) [inline yajsapi/executor/ctx.ts:299]: hmm.... can you separate dictionary values with semicolons indeed?

File: yajsapi/executor/index.ts [hunk 12]
```diff
@@ -208,40 +205,54 @@ export class Executor {
     payment_driver,
     payment_network,
     event_consumer,
-    network_address
+    network_address,
   }: ExecutorOpts) {
     this._subnet = subnet_tag ? subnet_tag : DEFAULT_SUBNET;
-    this._payment_driver = payment_driver ? payment_driver.toLowerCase() : (driver ? driver.toLowerCase() : DEFAULT_DRIVER);
-    this._payment_network = payment_network ? payment_network.toLowerCase() : (network ? network.toLowerCase() : DEFAULT_NETWORK);
+    this._payment_driver = payment_driver
```
Comment (shadeofblue) [inline yajsapi/executor/index.ts:211]: hmm... if we're splitting that into multiple lines, wouldn't a regular if/else syntax make more sense?

Comment (shadeofblue) [reply yajsapi/executor/index.ts:211]: plus, then, the deprecation warning could be included in the if/else and so overall, it would result in even less condition branches...

File: yajsapi/executor/index.ts [hunk 13]
```diff
@@ -589,28 +581,28 @@ export class Executor {
...
-              "Overriding batch timeout set with commit({timeout:...}) by value from ExecOptions"
-            );
+            logger.warn("Overriding batch timeout set with commit({timeout:...}) by value from ExecOptions");
           } else {
             exec_options.batch_timeout = batch.timeout();
           }
         }
-        const batch_deadline
-          = exec_options.batch_timeout ? ((Date.now() + exec_options.batch_timeout) / 1000) : undefined;
+        const batch_deadline = exec_options.batch_timeout
```
Comment (shadeofblue) [inline yajsapi/executor/index.ts:594]: again, if we're splitting the ternary into multiple lines, shouldn't we replace it with a more conventional if/else?

File: yajsapi/executor/strategy.ts [hunk 14]
```diff
@@ -132,7 +115,7 @@ export class LeastExpensiveLinearPayuMS {
 
     // The higher the expected price value, the lower the score.
     // The score is always lower than SCORE_TRUSTED and is always higher than 0.
-    const score: number = (SCORE_TRUSTED * 1.0) / (expected_price + 1.01);
+    const score: number = SCORE_TRUSTED / (expected_price + 1.01);
```
Comment (shadeofblue) [inline yajsapi/executor/strategy.ts:118]: isn't the cast to float needed here?

File: yajsapi/network/index.ts [hunk 15]
```diff
@@ -21,6 +29,11 @@ export class NetworkNode {
     };
   }
 
+  /**
```
Comment (shadeofblue) [inline yajsapi/network/index.ts:32]: just a comment: I guess docstrings et al could go into a separate pull request...

File: yajsapi/utils/promisify.ts [hunk 16]
```diff
@@ -1,4 +1,4 @@
-export default function promisify(fn: Function): Function {
+export default function promisify(fn: (...args) => void): (arg) => Promise<any> {
```
Comment (shadeofblue) [inline yajsapi/utils/promisify.ts:1]: I think I don't get the type definition for `fn` here -> doesn't it limit the functions passed here to ones that don't return any values? is that what is wanted?

**Review by shadeofblue — APPROVED** [review]
I guess I didn't find anything really out of place here ... but given the length of the pull request, I may have missed something... 

otherwise, I added some comments and questions...

File: examples/hello-world/hello.js (hunk 2 above)
Comment (mgordel) [reply examples/hello-world/hello.js:23]: this kind of formatting has been corrected automatically by the `prettier` tool with default rules. This can be tweaked of course, but I would leave the default rules which are often used in the js community

File: examples/low-level-api/listOffers.ts (hunk 3 above)
Comment (mgordel) [reply examples/low-level-api/listOffers.ts:5]: this kind of formatting has been corrected automatically by the `prettier` tool with default rules. This can be tweaked of course, but I would leave the default rules which are often used in the js community

File: examples/low-level-api/listOffers.ts (hunk 4 above)
Comment (mgordel) [reply examples/low-level-api/listOffers.ts:17]: yeah .. JS can be weird sometimes. In JS, `const` means that the identifier can’t be assigned, not that values is immutable. And linter show errors if you use `let` for identifiers which are not reassigned.

File: examples/unsubscribe.ts (hunk 5 above)
Comment (mgordel) [reply examples/unsubscribe.ts:4]: yes, you are right, i'll fix it.

File: tests/goth/test_run_blender.py (hunk 6 above)
Comment (mgordel) [reply tests/goth/test_run_blender.py:106]: in the first one we use node directly so we can only execute js file. we can't run ts file (typescript) in node engine. First we need to compile ts -> js. To avoid this in this test example (i wanted to test ts version in goth), we can use the ts-node tool, which compiles and runs the ts file on the fly. It is hidden under the `ts:blender` command in package.json, so if we want to run more advanced scripts than node {js_file} we need to define it in package.json and use npm, something like poetry and poe tool..

File: tests/utils/promisify.ts (hunk 7 above)
Comment (mgordel) [reply tests/utils/promisify.ts:21]: this is TS syntax. The error variable can be undefined (by default type is Error | undefined), in which case referring to the message property will throw error. If we don't define an if statement that checks if error is undefined, the typescript compiler will report an error. The syntax `error!.message` is such a hack that tells TS compiler that we know that error var will definitely not be undefined, and ts compiler not show error in this case.

File: yajsapi/executor/ctx.ts (hunk 8 above)
Comment (mgordel) [reply yajsapi/executor/ctx.ts:44]: Yes exactly

File: yajsapi/executor/ctx.ts (hunk 9 above)
Comment (mgordel) [reply yajsapi/executor/ctx.ts]: Yes, i agree

File: yajsapi/executor/ctx.ts (hunk 11 above)
Comment (mgordel) [reply yajsapi/executor/ctx.ts:301]: ts suggests using `type` instead of an object (dictionary).
the `type` construct uses semicolons. https://www.typescriptlang.org/docs/handbook/2/objects.html

File: yajsapi/executor/index.ts (hunk 12 above)
Comment (mgordel) [reply yajsapi/executor/index.ts:211]: yes, i agree

File: yajsapi/utils/promisify.ts (hunk 16 above)
Comment (mgordel) [reply yajsapi/utils/promisify.ts:1]: yes, it should be `(...args) => any`

File: examples/low-level-api/listOffers.ts (hunk 3 above)
Comment (mgordel) [reply examples/low-level-api/listOffers.ts:5]: yes, you are right, i'll fix it.
